### PR TITLE
Adding Cycling Status option

### LINKED ISSUE #324: Multiple Now Playing Messages
I'd like to put something like 10 now playing messages in the config, and have the bot switch once a hour or so.

### PR DESCRIPTION
**Pull request by underscore11code**
This adds the option for multiple "**Playing:** XYZ" statuses that get cycled through at a customizable rate, minimum 1 minute/update.

Also closes #324 and #526* (see comment at bottom).

This is currently lacking translations, unless you want me to google translate it. However, config changes have been done in such a way to maintain backwards compatibility.

Edit by Vankka: 
\* This does not _fully_ resolve 526, as this does not include Listening to, Watching or Streaming

### CODE REVIEW

**Comment by Vankka** [comment]
This pull request does *not* address #526

**Comment by Vankka** [comment]
Config options & comments, are missing/ inaccuret (they could also follow the principals of config files)

**Comment by underscore11code** [comment]
> This pull request does not address #526

The status gets updated even if it is not cycling. Forgot to mention that in initial message, my apologies.

**Comment by Vankka** [comment]
*Sorry, it does somewhat adress it, just doesn't address Watching/Listening to/whatever

**Comment by underscore11code** [comment]
Rewritten, should be much cleaner now

File: src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class PlayingStatusUpdater extends Thread {
+    public PlayingStatusUpdater() {
+        setName("DiscordSRV - Status Updater");
+    }
+
+    private int lastStatus=0;
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java]: Add a couple spaces

File: src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    @Override
+    public void run() {
+        while (true) {
+            int rate;
+            //Breaking change to config
+            try {
+                rate=DiscordSRV.config().getInt("StatusUpdateRateInMinutes");
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java]: A couple spaces

File: src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            } catch(IllegalArgumentException e){
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java]: A coupke spaces

File: src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    public PlayingStatusUpdater() {
+        setName("DiscordSRV - Status Updater");
+    }
+
+    private int lastStatus=0;
+    @Override
+    public void run() {
+        while (true) {
+            int rate;
+            //Breaking change to config
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java]: Add a space

File: src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                    statuses.add(dynamic.convert().intoString());
+                }
+                String status = statuses.get(lastStatus);
+
+                // Increment and wrap around
+                lastStatus++;
+                if (lastStatus == statuses.size() - 1)
+                    lastStatus = 0;
+
+                if(!StringUtils.isEmpty(status))
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java]: Add a space

File: src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    @Override
+    public void run() {
+        while (true) {
+            int rate;
+            // Breaking change to config
+            try {
+                rate = DiscordSRV.config().getInt("StatusUpdateRateInMinutes");
+            } catch (IllegalArgumentException e) {
+                DiscordSRV.error("\"StatusUpdaterRateInMinutes\" not found in config");
+                rate=2;
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java]: Missed this before, style

File: src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            } catch (InterruptedException e) {
+                DiscordSRV.error("Broke from Status Updater thread: sleep interrupted");
+                return;
+            }
+        }
+    }
+
+    @Getter private static File playerDataFolder = new File(Bukkit.getWorlds().get(0).getWorldFolder().getAbsolutePath(), "/playerdata");
+    @SuppressWarnings({"SpellCheckingInspection", "ConstantConditions"})
+    private static String applyPlaceholders(String input) {
```
Comment (Vankka) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PlayingStatusUpdater.java]: This is a direct copy of https://github.com/DiscordSRV/DiscordSRV/blob/75d61bc51d522e2b85320253dfc60bac0586d2d2/src/main/java/github/scarsz/discordsrv/objects/threads/ChannelTopicUpdater.java#L67 (should be 1 method, not 2. Bad practice)

**Comment by underscore11code** [comment]
Style issues should be resolved, Ill move the duplicate method to somewhere in util I guess

**Comment by underscore11code** [comment]
Duplicate method resolved, apparently it already was parsing through PlaceholderAPI, just was never useful due to the lack of updates. https://github.com/DiscordSRV/DiscordSRV/blob/75d61bc51d522e2b85320253dfc60bac0586d2d2/src/main/java/github/scarsz/discordsrv/util/DiscordUtil.java#L438-L456

**Comment by underscore11code** [comment]
For the sake of not making a class dedicated to the single placeholder parsing method, or sticking it into another file like LangUtil where it doesn't really fit, I have not implemented the built-in placeholders. 

The class has been renamed as requested to PresenceUpdater.

File: src/main/java/github/scarsz/discordsrv/objects/threads/PresenceUpdater.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    private int lastStatus = 0;
+    @Override
+    public void run() {
+        while (true) {
+            int rate;
+            // Breaking change to config
+            try {
+                rate = DiscordSRV.config().getInt("StatusUpdateRateInMinutes");
```
Comment (pkrok01) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PresenceUpdater.java]: You misspelled `StatusUpdaterRateInMinutes`, which causes it to error.

File: src/main/java/github/scarsz/discordsrv/objects/threads/PresenceUpdater.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+                // Increment and wrap around
+                lastStatus++;
+                if (lastStatus == statuses.size() - 1)
+                    lastStatus = 0;
+
+                if (!StringUtils.isEmpty(status))
+                    DiscordUtil.setGameStatus(status);
+                else
+                    DiscordSRV.debug("Skipping status update cycle, status was null");
```
Comment (pkrok01) [inline src/main/java/github/scarsz/discordsrv/objects/threads/PresenceUpdater.java]: Doesn't actually rotate status. My tests show it sticks to the first option no matter what.

**Comment by underscore11code** [comment]
Fixed and tested, works now.

**Pull request merged**